### ISSUE TITLE
META Issue: Packages should be cryptographically signed

### ISSUE THREAD

**Issue by scunningham** [elastic/package-spec#46]
This may be covered elsewhere, but I'll add just in case.

The packages contains instructions that could significantly modify the structure of a customer's indices; ILM policy, transforms, mappings etc.  We should not be reliant solely on HTTPS DNS name validation to prove provenance of the package.  A determined state level attacker could forge a certificate and play man in the middle on any of the requests, serving up packages intended to disrupt the target.

This will become a larger issue as we add more content to the packages.  I am thinking, for example, of SIEM ruleset for Elastic Security.  In that case, an attacker could undermined a target's security posture by replacing or modifying effective detection rules.

The packages should be signed much like the code we deliver to customers.  The recipient, in this case Kibana, should validate the signature before installation.  To support non-elastic or dev packages, the UI should warn if a package is not signed, but allow the customer to override.

- [x] https://github.com/elastic/package-registry/issues/728

**Comment by ruflin** [comment]
Adding https://github.com/elastic/package-registry/issues/168 here to link some previous discussions.

**Comment by jlind23** [comment]
After a discussion with @mtojek - We need to add some fields inside package manifest to ensure that our packages are signed and you can trust them.

**Comment by mtojek** [comment]
Let me sketch the MVP.

I based my research on the GPG signing workflow: https://www.gnupg.org/gph/en/manual/x135.html . The biggest inconvenience is the archiving logic which creates .zip archives on the fly. This will change if we decide to switch to store only .zip files in the Package Storage. 

1. Only Elastic can sign packages.
2. Private keys are internal and can’t escape the CI.
3. While building the release we have to generate <package-name>-<package-version>.sig files in the CI environment.
4. Public key is stored in the Package Storage and distributed together with its Docker images.
5. Signature files (.sig) are stored in the signatures/ directory in Package Storage.
6. Package Registry exposes additional property “signature”.
7. After Kibana downloads the .zip package it verifies the package signature.

Comments:
1. Package .zip archives are created dynamically, which means that the file content may slightly differ if we change the Go version or adjust the archive serving logic. We can acknowledge this as it will be fixed when we switch to static .zip files instead of the dynamic option.
2. Another option would be adopting a different signature algorithm, similar to content addressing: https://docs.ipfs.io/concepts/content-addressing/
signature = hash(signature(f1) + signature(f2) + … + signature(fn)), where fn are alphanum sorted files in the archive.
2.1 Kibana must be aware of this signing algorithm
2.2 It isn’t GPG anymore
2.3 We can change implementation/compression rate of archiver

Let me know what you think.

cc @jsoriano

**Comment by mtojek** [comment]
I described the workflow in: https://github.com/elastic/package-registry/issues/728#issuecomment-933463101

**Comment by mtojek** [comment]
Based on the comment in https://github.com/elastic/package-registry/issues/728#issuecomment-947515518 , we agreed that we don't need changes in package-spec, so let the EPR's issue be the meta for other projects.

**Comment by jlind23** [comment]
Reopened as even if proposal is settled, nothing is production ready yet.

**Comment by alvarolobato** [comment]
@jlind23 this is still planned for 8.0 isn't it?

**Comment by mtojek** [comment]
No, this will be enabled since Day 1 of the improved Package Storage (one of future minor 8.x). We need changes in Kibana and tooling. Package Registry as the application is the only one ready to support signatures.

**Comment by alvarolobato** [comment]
ok, thanks @mtojek 
@mukeshelastic do we have a plan for the rest of this work to be done. There's concerns about delaying this too much.

**Comment by mukeshelastic** [comment]
Adding @akshay-saraswat who worked on the requirements and UX for package signing..

@mtojek I am assuming that moving the package storage from a folder structure to a zip file is a pre-req for signing it. 

@akshay-saraswat @jlind23 is there anything else on the team's plate that we can drop to help focus on the storage and package signing part?

**Comment by mtojek** [comment]
> @mtojek I am assuming that moving the package storage from a folder structure to a zip file is a pre-req for signing it.

Precisely, it's a must-have.

The current form of Package Storage stores only unpacked files and shouldn't store any ZIP files as it's a Git repository. It's strongly not recommended. Hence we decided to introduce it when we have the storage ready (see the design document I sent a few days ago). All other approaches we can take will be workarounds or hacks, which won't guarantee security level (like signing packages on the fly, during/before building Docker images).

From our perspective, we are on track and following the plan we agreed on (milestones).

> @akshay-saraswat @jlind23 is there anything else on the team's plate that we can drop to help focus on the storage and package signing part?

We are aware of the team capacity which usually is 1 person-day (considering the 2nd person present and working on daily operations), since holidays until mid-Jan it's 0. Even though, In this case, it is mainly a matter of dependencies and requirements for particular project components.
